File: app/src/main/java/de/bahnhoefe/deutschlands/bahnhofsfotos/util/BitmapCache.java
package de.bahnhoefe.deutschlands.bahnhofsfotos.util;

import android.graphics.Bitmap;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;

/**
 * A cache for station photos.
 */
public class BitmapCache {
    private final String TAG = BitmapCache.class.getSimpleName();

    /**
     * The singleton.
     */
    private static BitmapCache instance = null;

    /**
     * Get the BitmapCache instance.
     *
     * @return the single instance of BitmapCache.
     */
    public static BitmapCache getInstance() {
        synchronized (BitmapCache.class) {
            if (instance == null) {
                instance = new BitmapCache();
            }
            return instance;
        }
    }

    private BitmapCache() {
        cache = new HashMap<>(10);
        requests = new HashMap<>(3);
    }

    private final HashMap<URL, Collection<BitmapAvailableHandler>> requests;

    private final HashMap<URL, Bitmap> cache;

    private class Request implements BitmapAvailableHandler {
        /**
         * This gets called if the requested bitmap is available. Write Bitmap to cache, find the requestor and pass on the result.
         *
         * @param bitmap the fetched Bitmap for the notification. May be null
         */
        @Override
        public void onBitmapAvailable(@Nullable Bitmap bitmap, URL url) {
            // save the image in the cache if it was constructed
            if (bitmap != null) {
                cache.put(url, bitmap);
            }

            // inform all requestors about the available image
            synchronized (requests) {
                var handlers = requests.remove(url);
                if (handlers == null) {
                    Log.wtf(TAG, "Request result without a saved requestor. This should never happen.");
                } else {
                    handlers.forEach(handler -> handler.onBitmapAvailable(bitmap, url));
                }
            }
        }
    }

    /**
     * Get a picture for the given URL, either from cache or by downloading.
     * The fetching happens asynchronously. When finished, the provided callback interface is called.
     *
     * @param callback    the BitmapAvailableHandler to call on completion.
     * @param resourceUrl the URL to fetch
     */
    public void getPhoto(BitmapAvailableHandler callback, @NonNull String resourceUrl) {
        URL url = null;
        try {
            url = new URL(resourceUrl);
            getPhoto(callback, url);
        } catch (MalformedURLException e) {
            Log.e(TAG, "Couldn't load photo from malformed URL " + resourceUrl);
            callback.onBitmapAvailable(null, url);
        }
    }

    /**
     * Get a picture for the given URL, either from cache or by downloading.
     * The fetching happens asynchronously. When finished, the provided callback interface is called.
     *
     * @param callback    the BitmapAvailableHandler to call on completion.
     * @param resourceUrl the URL to fetch
     */
    public void getPhoto(BitmapAvailableHandler callback, @NonNull URL resourceUrl) {
        var bitmap = cache.get(resourceUrl);
        if (bitmap == null) {
            var downloader = new BitmapDownloader(new Request(), resourceUrl);
            synchronized (requests) {
                var handlers = requests.get(resourceUrl);
                if (handlers == null) {
                    handlers = new ArrayList<>();
                    handlers.add(callback);
                    requests.put(resourceUrl, handlers);
                } else {
                    handlers.add(callback);
                }
            }
            downloader.start();
        } else {
            callback.onBitmapAvailable(bitmap, resourceUrl);
        }

    }

}

File: app/src/main/kotlin/de/bahnhoefe/deutschlands/bahnhofsfotos/util/BitmapCache.kt
package de.bahnhoefe.deutschlands.bahnhofsfotos.util

import android.graphics.Bitmap
import android.util.Log
import java.net.MalformedURLException
import java.net.URI
import java.util.function.Consumer

private val TAG = BitmapCache::class.java.simpleName

/**
 * A cache for station photos.
 */
object BitmapCache {

    private val requests = mutableMapOf<URI, MutableCollection<BitmapAvailableHandler>>()
    private val cache = mutableMapOf<URI, Bitmap>()

    /**
     * Get a picture for the given URL, either from cache or by downloading.
     * The fetching happens asynchronously. When finished, the provided callback interface is called.
     *
     * @param resourceUri the URI to fetch
     * @param callback    the BitmapAvailableHandler to call on completion.
     */
    fun getPhoto(resourceUri: String, callback: BitmapAvailableHandler) {
        try {
            getPhoto(URI.create(resourceUri), callback)
        } catch (_: MalformedURLException) {
            Log.e(TAG, "Couldn't load photo from malformed URL $resourceUri")
            callback.onBitmapAvailable(null)
        }
    }

    /**
     * Get a picture for the given URL, either from cache or by downloading.
     * The fetching happens asynchronously. When finished, the provided callback interface is called.
     *
     * @param resourceUrl the URL to fetch
     * @param callback    the BitmapAvailableHandler to call on completion.
     */
    private fun getPhoto(resourceUrl: URI, callback: BitmapAvailableHandler) {
        val bitmap = cache[resourceUrl]
        if (bitmap == null) {
            val downloader = BitmapDownloader(resourceUrl) { fetchedBitmap: Bitmap? ->
                if (fetchedBitmap != null) {
                    cache[resourceUrl] = fetchedBitmap
                }

                // inform all requesters about the available image
                synchronized(requests) {
                    requests.remove(resourceUrl)
                        ?.forEach(Consumer { handler ->
                            handler.onBitmapAvailable(fetchedBitmap)
                        })
                        ?: Log.e(
                            TAG,
                            "Request result without a saved requester. This should never happen."
                        )
                }
            }
            synchronized(requests) {
                var handlers = requests[resourceUrl]
                if (handlers == null) {
                    handlers = ArrayList()
                    handlers.add(callback)
                    requests.put(resourceUrl, handlers)
                } else {
                    handlers.add(callback)
                }
            }
            downloader.start()
        } else {
            callback.onBitmapAvailable(bitmap)
        }
    }
}
